### PR TITLE
Cfspdk 1075 button press cb

### PR DESCRIPTION
**Pull request by YHusain1**
This adds the functionality of a callback for user button presses, and gives information on the timestamp, duration and type of button press.

### CODE REVIEW

File: samples/button_press_cb/src/button.c [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+static void button_press_handler(const struct device *port, struct gpio_callback *cb,
+				 gpio_port_pins_t pins)
+{
+	k_timer_stop(&debounce_timer);
+	k_timer_start(&debounce_timer, K_MSEC(50), K_NO_WAIT);
+	button_state = BUTTON_CB_STATE_DEBOUNCING;
+}
+
+/* debounce timer handler */
+static void button_debounce_timer_hanlder(struct k_timer *debounce_timer)
```
Comment (johnlange2) [inline samples/button_press_cb/src/button.c:36]: please correct spelling of handler

**Review by jtbaumann — APPROVED** [review]

File: samples/button_press_cb/README.txt [hunk 2]
```diff
@@ -0,0 +1 @@
+TODO
```
Comment (johnlange2) [inline samples/button_press_cb/README.txt:1]: Need to write some basic content here, at a minimum including how to build

**Comment by YHusain1** [comment]
@raamratt2 
can you help review this PR.
@johnlange2 mentioned I need to add the testing instructions to make it easy for your team. Can you read the README file and see if it is per what you expect? or maybe guide me to what I need to change to have it per QA expectations

File: samples/button_press_cb/src/button.c [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+		last_button_press_time = save_last_button_press;
+		/* Invoke the registered user callback handler */
+		registered_callback(last_button_press_time, button_press_duration,
+				    BUTTON_PRESS_TIMEOUT);
+	}
+	k_timer_stop(p_timeout_timer);
+}
+
+/* debounce timer handler */
+static void button_debounce_timer_handler(struct k_timer *p_debounce_timer)
```
Comment (jtbaumann) [inline samples/button_press_cb/src/button.c:55]: Why are we implementing our own debounce? We should be using the built in debounce mechanism (gpio keys)

**Comment by raamratt2** [comment]
> @raamratt2 can you help review this PR. @johnlange2 mentioned I need to add the testing instructions to make it easy for your team. Can you read the README file and see if it is per what you expect? or maybe guide me to what I need to change to have it per QA expectations

I think it would be important in the readme file to include what you had described in your comments under button.h file (line 26-32). It would help explain in the readme the unit of measurement for each variable we're expected. I would have no grasp as to what the numbers mean for timestamp and duration.

**Comment by YHusain1** [comment]
> > @raamratt2 can you help review this PR. @johnlange2 mentioned I need to add the testing instructions to make it easy for your team. Can you read the README file and see if it is per what you expect? or maybe guide me to what I need to change to have it per QA expectations
> 
> I think it would be important in the readme file to include what you had described in your comments under button.h file (line 26-32). It would help explain in the readme the unit of measurement for each variable we're expected. I would have no grasp as to what the numbers mean for timestamp and duration.

I have updated the README file with your suggestions. I will resolve this comment and move on if that's okay with you. @raamratt2

File: samples/button_press_cb/src/button.c (hunk 3 above)
Comment (YHusain1) [reply samples/button_press_cb/src/button.c:55]: I did not know we have that one.
As for conforming to the acceptance criteria in the ticket, I believe having 'some; kind of debounce is good for now.
I would love to explore the already existing debounce mechanism and integrate it to this sample, but it's too late now for this to be done today, we need to release this for tomorrow's 1.20

**Comment by raamratt2** [comment]
> > > @raamratt2 can you help review this PR. @johnlange2 mentioned I need to add the testing instructions to make it easy for your team. Can you read the README file and see if it is per what you expect? or maybe guide me to what I need to change to have it per QA expectations
> > 
> > 
> > I think it would be important in the readme file to include what you had described in your comments under button.h file (line 26-32). It would help explain in the readme the unit of measurement for each variable we're expected. I would have no grasp as to what the numbers mean for timestamp and duration.
> 
> I have updated the README file with your suggestions. I will resolve this comment and move on if that's okay with you. @raamratt2

looks good to me this comment can be resolved.

**Review by johnlange2 — APPROVED** [review]

**Review by raamratt2 — APPROVED** [review]

**Pull request merged**